src/libs/ck-libs/allGather/allGather.ci

```diff
@@ -0,0 +1,59 @@
+module allGather
+{
+  include "vector";
+
+  message allGatherMsg;
+
+  group AllGather
+  {
+    entry AllGather(int size, int type, int seed);
+    entry[reductiontarget] void startGather();
+    entry void recvRing(int sender, CkNcpyBuffer data);
+    entry[nokeep] void local_buff_done(CkDataMsg * m);
+    entry void Hypercube()
+    {
+      for (iter = 0; iter < numHypercubeIter; iter++)
+      {
+        serial
+        {
+          if (HypercubeRecursiveDoubling)
+            HypercubeToSend = (n + CkMyPe() - (int)pow(2, iter)) % n;
+          else
+            HypercubeToSend = CkMyPe() ^ ((int)pow(2, iter));
+        }
+        serial
+        {
+          if (HypercubeRecursiveDoubling && iter == numHypercubeIter - 1)
+          {
+            std::vector<CkNcpyBuffer> hyperCubeStoreCopy(
+                hyperCubeStore.begin(), hyperCubeStore.begin() + (n - (int)pow(2, iter)));
+            std::vector<int> hyperCubeIndxCopy(
+                hyperCubeIndx.begin(), hyperCubeIndx.begin() + (n - (int)pow(2, iter)));
+            thisProxy[HypercubeToSend].recvHypercube(iter, hyperCubeStoreCopy,
+                                                     hyperCubeIndxCopy);
+          }
+          else
+            thisProxy[HypercubeToSend].recvHypercube(iter, hyperCubeStore, hyperCubeIndx);
+        }
+        when recvHypercube[iter](int ref, std::vector<CkNcpyBuffer> data,
```

Comment:
It would be more efficient to send these vector's as C-style arrays to avoid copying from the Charm++ msg to a std::vector on the receive side. On the sender-side you can keep them as std::vector's and pass the .size() and .data() separately.